File: src/frontend/recruti_front/src/app/services/offer.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Offer } from '../models/offer.model';
import { PagedResponse } from '../models/paged-response.model';

export interface OfferRequest {
  titre: string;
  description: string;
  tags: string;
  available: boolean;
  idRecruteur: number;
  dateEmission: string;
}

export interface OfferResponse {
  id: number;
  dateEmission: string;
  titre: string;
  description: string;
  tags: string;
  available: boolean;
  idRecruteur: number;
}

@Injectable({
  providedIn: 'root'
})
export class OfferService {
  private readonly baseUrl = `${environment.apiUrl}/api/offers`;

  constructor(private readonly http: HttpClient) {}

  /**
   * Récupère les offres disponibles (paginé avec filtre)
   */
  getAvailableOffers(page = 1, limit = 6, titre = ''): Observable<PagedResponse<Offer>> {
    const params = new HttpParams()
      .set('page', page)
      .set('limit', limit)
      .set('titre', titre);
    return this.http.get<PagedResponse<Offer>>(`${this.baseUrl}/available`, { params });
  }

  /**
   * Récupère toutes les offres
   */
  getAllOffers(): Observable<Offer[]> {
    return this.http.get<Offer[]>(this.baseUrl);
  }

  /**
   * Récupère les offres (paginé)
   */
  getOffers(page = 1, limit = 6): Observable<PagedResponse<Offer>> {
    const params = new HttpParams()
      .set('page', page)
      .set('limit', limit);
    return this.http.get<PagedResponse<Offer>>(`${this.baseUrl}`, { params });
  }

  /**
   * Récupère toutes les offres d'un recruteur spécifique (paginé)
   * GET /api/offers/recruiter/{recruiterId}?page=1&limit=6
   */
  getMyOffers(recruiterId: number, page: number = 1, limit: number = 6): Observable<PagedResponse<OfferResponse>> {
    const params = new HttpParams()
      .set('page', page)
      .set('limit', limit);
    return this.http.get<PagedResponse<OfferResponse>>(
      `${this.baseUrl}/recruiter/${recruiterId}`,
      { params }
    );
  }

  /**
   * Crée une nouvelle offre
   * POST /api/offers
   */
  addOffer(offer: OfferRequest): Observable<Offer> {
    return this.http.post<Offer>(this.baseUrl, offer);
  }

  /**
   * Met à jour une offre existante
   * PUT /api/offers/{id}
   */
  updateOffer(id: number, offer: OfferRequest): Observable<Offer> {
    return this.http.put<Offer>(`${this.baseUrl}/${id}`, offer);
  }

  /**
   * Supprime une offre
   * DELETE /api/offers/{id}
   */
  deleteOffer(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: src/backend/recruti_backend/src/main/java/tn/recruti/recruti_backend/service/OfferService.java
package tn.recruti.recruti_backend.service;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import tn.recruti.recruti_backend.dto.OfferRequestDto;
import tn.recruti.recruti_backend.dto.OfferResponseDto;
import tn.recruti.recruti_backend.dto.PagedResponse;
import tn.recruti.recruti_backend.Exception.DuplicateResourceException;
import tn.recruti.recruti_backend.Exception.RessourceNotFoundException;
import tn.recruti.recruti_backend.model.Offer;
import tn.recruti.recruti_backend.model.User;
import tn.recruti.recruti_backend.repository.OfferRepository;
import tn.recruti.recruti_backend.repository.UserRepository;

@Service
@RequiredArgsConstructor
public class OfferService {

    private final OfferRepository offerRepository;
    private final UserRepository userRepository;

   // MODIFIER la méthode addOffer() pour ajouter la validation

public Offer addOffer(OfferRequestDto dto) {
    User recruteur = userRepository.findById(dto.idRecruteur())
        .orElseThrow(() -> new RessourceNotFoundException(
            "Recruteur not found: " + dto.idRecruteur()));

    // Vérifier si une offre avec le même titre existe déjà pour ce recruteur
    if (offerRepository.existsByTitreAndRecruteurId(dto.titre(), dto.idRecruteur())) {
        throw new DuplicateResourceException(
            "Une offre avec le titre '" + dto.titre() + "' existe déjà pour ce recruteur");
    }

    Offer offer = new Offer();
    offer.setDateEmission(dto.dateEmission());
    offer.setTitre(dto.titre());
    offer.setDescription(dto.description());
    offer.setTags(dto.tags());
    offer.setAvailable(dto.available());
    offer.setRecruteur(recruteur);

    return offerRepository.save(offer);
}
    // ── Update ───────────────────────────────────────────
    public Offer updateOffer(Long id, OfferRequestDto dto) {
        Offer offer = offerRepository.findById(id)
            .orElseThrow(() -> new RessourceNotFoundException("Offer not found: " + id));

        User recruteur = userRepository.findById(dto.idRecruteur())
            .orElseThrow(() -> new RessourceNotFoundException(
                "Recruteur not found: " + dto.idRecruteur()));

        offer.setDateEmission(dto.dateEmission());
        offer.setTitre(dto.titre());
        offer.setDescription(dto.description());
        offer.setTags(dto.tags());
        offer.setAvailable(dto.available());
        offer.setRecruteur(recruteur);

        return offerRepository.save(offer);
    }

    // ── Get all (paginé) ──────────────────────────────────

    // Dans OfferService.java
private OfferResponseDto toDto(Offer offer) {
    return new OfferResponseDto(
        offer.getId(),
        offer.getDateEmission(),
        offer.getTitre(),
        offer.getDescription(),
        offer.getTags(),
        offer.isAvailable(),
        offer.getRecruteur() != null ? offer.getRecruteur().getId() : null
    );
}

public PagedResponse<OfferResponseDto> getAllOffers(int page, int limit) {
    Pageable pageable = PageRequest.of(page - 1, limit,
        Sort.by("dateEmission").descending());
    Page<Offer> result = offerRepository.findAll(pageable);

    return new PagedResponse<>(
        result.getContent().stream().map(this::toDto).toList(),
        page,
        limit,
        result.getTotalElements(),
        result.getTotalPages()
    );
}

// Ajouter cette méthode après getAllOffers()

/**
 * Récupère toutes les offres d'un recruteur spécifique (paginé)
 * GET /api/offers/recruiter/{recruiterId}
 */
public PagedResponse<OfferResponseDto> getOffersByRecruiterId(Long recruiterId, int page, int limit) {
    // Vérifier que le recruteur existe
    if (!userRepository.existsById(recruiterId)) {
        throw new RessourceNotFoundException("Recruteur not found: " + recruiterId);
    }
    
    Pageable pageable = PageRequest.of(page - 1, limit,
        Sort.by("dateEmission").descending());
    Page<Offer> result = offerRepository.findByRecruteurId(recruiterId, pageable);

    return new PagedResponse<>(
        result.getContent().stream().map(this::toDto).toList(),
        page,
        limit,
        result.getTotalElements(),
        result.getTotalPages()
    );
}
   // ── Get by id ─────────────────────────────────────────
    public Offer getOfferById(Long id) {
        return offerRepository.findById(id)
            .orElseThrow(() -> new RessourceNotFoundException("Offer not found: " + id));
    }

    // ── Delete ────────────────────────────────────────────
    public void deleteOffer(Long id) {
        if (!offerRepository.existsById(id)) {
            throw new RessourceNotFoundException("Offer not found: " + id);
        }
        offerRepository.deleteById(id);
    }
}